flux-middle/src/rustc/mir.rs

```diff
@@ -40,21 +41,35 @@ pub struct LocalDecl {
     pub source_info: SourceInfo,
 }
 
-pub struct Terminator {
-    pub kind: TerminatorKind,
+pub struct Terminator<'tcx> {
+    pub kind: TerminatorKind<'tcx>,
     pub source_info: SourceInfo,
 }
 
 #[derive(Debug)]
-pub enum TerminatorKind {
+pub struct CallSubsts<'tcx> {
+    pub orig: SubstsRef<'tcx>,
+    pub lowered: List<GenericArg>,
+}
+
+/// An `Instance` is the resolved call-target at a particular trait-call-site
+#[derive(Debug)]
+pub struct Instance {
+    pub impl_f: DefId,
+    pub substs: List<GenericArg>,
+}
+
+#[derive(Debug)]
+pub enum TerminatorKind<'tcx> {
     Return,
     Call {
         func: DefId,
-        substs: List<GenericArg>,
+        substs: CallSubsts<'tcx>,
         args: Vec<Operand>,
         destination: Place,
         target: Option<BasicBlock>,
         cleanup: Option<BasicBlock>,
+        instance: Option<Instance>,
```

Comment:
I wonder if it is cleaner to just store the most specific "Instance" (DefId/Substs pair) we can find here instead of both, then the rest of the checking will proceed as normal. But at the same time, I guess having more information is better than having less information 🤔

Reply:
Yes, unclear to me too, but didn't want to erase the original info...